### PR TITLE
Performance optimizations

### LINKED ISSUE #520: notebook describing op.apply
We need a jupyter notebook explaining, perhaps with stupid examples, the mechanics of ``op.apply``. in particular:
* what it does:
** potentially triggers compilation
** potentially allocates data
**generation of default runtime values
** execution of code
* accepted parameters (semantics of *_m, *_M, *_n, ...)
* what else?

### PR DESCRIPTION
**Pull request by FabioLuporini**
The following new optimizations are disabled by default -- until I'll have benchmarked them properly

* Autopadding of Function/TimeFunction
* Hierarchical loop blocking
* Loop rounding (see docstrings for explanation of what this is)

Also, ``omp collapse`` is now applied more aggressively than before (based on my experiments, it generally produces faster code)

This PR also ships a jupyter notebook describing ``op.apply``

fixes #520

### CODE REVIEW

File: devito/ir/support/basic.py [hunk 1]
```diff
@@ -545,6 +549,24 @@ def is_inplace(self, dim=None):
         """Stronger than ``is_indep()``, as it also compares the timestamps."""
         return self.source.lex_eq(self.sink) and self.is_indep(dim)
 
+    @memoized_meth
+    def is_storage_volatile(self, dims=None):
+        """
+        True if a storage-volatile dependence, False otheriwse.
```
Comment (georgebisbas) [inline devito/ir/support/basic.py]: otherwise

File: devito/dle/transformer.py [hunk 2]
```diff
@@ -2,14 +2,14 @@
...
-from devito.dle.rewriters import CPU64Rewriter, CustomRewriter, State
+from devito.dle.rewriters import CPU64Rewriter, CustomRewriter, PlatformRewriter, State
 from devito.logger import dle as log, dle_warning as warning
 from devito.parameters import configuration
 
 __all__ = ['dle_registry', 'modes', 'transform']
 
 
-dle_registry = ('advanced', 'speculative')
+dle_registry = ('noop', 'advanced', 'speculative')
```
Comment (mloubout) [inline devito/dle/transformer.py]: Wait I thought there was a basic here at some points, examples may need update from this

File: devito/types/dense.py [hunk 3]
```diff
@@ -50,17 +51,19 @@ class DiscreteFunction(AbstractCachedFunction, ArgProvider):
...
             # Staggering metadata
             self._staggered = self.__staggered_setup__(**kwargs)
 
+            # Now that *all* __X_setup__ hooks have been called, we can let the
+            # superclass constructor do its job
+            super(DiscreteFunction, self).__init__(*args, **kwargs)
+
+            # There may or may not be a `Grid` attached to the DiscreteFunction
+            self._grid = kwargs.get('grid')
+
```
Comment (mloubout) [inline devito/types/dense.py]: Why is the order changed?

File: devito/types/dense.py [hunk 4]
```diff
@@ -1004,12 +1010,35 @@ def __halo_setup__(self, **kwargs):
...
+                # * all first grid points along the `fvd` will be cache-aligned
+                # * there is enough room to round up the loop trip counts to maximize
+                #   the effectiveness SIMD vectorization
+                fvd_pad_size = (ub(self._size_nopad[fvd]) - self._size_nopad[fvd]) + vl
+                padding.append((0, fvd_pad_size))
+                return tuple(padding)
+            else:
+                return tuple((0, 0) for d in self.dimensions)
+        elif isinstance(padding, int):
+            return tuple((0, padding) if d.is_Space else (0, 0) for d in self.dimensions)
```
Comment (mloubout) [inline devito/types/dense.py]: What's the difference with the one in basic? Not sure this one is needed too.

File: devito/types/dense.py [hunk 5]
```diff
@@ -1296,6 +1327,10 @@ def __init__(self, *args, **kwargs):
             super(SubFunction, self).__init__(*args, **kwargs)
             self._parent = kwargs['parent']
 
+    def __padding_setup__(self, **kwargs):
+        # SubFunctions aren't expected to be used in time-consuming loops
+        return tuple((0, 0) for i in range(self.ndim))
```
Comment (mloubout) [inline devito/types/dense.py]: Not sure if true but I guess it's the case as of now

File: devito/yask/types.py [hunk 6]
```diff
@@ -71,6 +71,10 @@ def __new__(cls, *args, **kwargs):
             newobj = cls.__base__.__new__(klass, *args, **kwargs)
         return newobj
 
+    def __padding_setup__(self, **kwargs):
+        # YASK calculates the padding, so we bypass the dense.Function's autopadding
+        return tuple((0, 0) for i in range(self.ndim))
```
Comment (mloubout) [inline devito/yask/types.py]: Shouldn't it be a `pass` instead if ignored? or just `()`

File: tests/test_autotuner.py [hunk 7]
```diff
@@ -102,7 +102,7 @@ def test_timesteps_per_at_run():
 @switchconfig(profiling='advanced')
 def test_mode_runtime_forward():
     """Test autotuning in runtime mode."""
-    grid = Grid(shape=(64, 64, 64))
+    grid = Grid(shape=(96, 96, 96))
```
Comment (mloubout) [inline tests/test_autotuner.py]: Any reason why?

File: tests/test_autotuner.py [hunk 8]
```diff
@@ -181,7 +181,7 @@ def test_tti_aggressive():
     wave_solver = tti_operator(dse='aggressive', dle=('advanced', {'openmp': False}))
     op = wave_solver.op_fwd(kernel='centered', save=False)
     op.apply(time=0, autotune='aggressive')
-    assert op._state['autotuning'][0]['runs'] == 33
+    assert op._state['autotuning'][0]['runs'] == 28
```
Comment (mloubout) [inline tests/test_autotuner.py]: I assume this is because the change of block sizes?

File: tests/test_autotuner.py [hunk 9]
```diff
@@ -296,12 +296,33 @@ def test_multiple_blocking():
     assert len(op._state['autotuning'][0]['tuned']) == 5
 
 
+def test_hierarchical_blocking():
+    grid = Grid(shape=(64, 64, 64))
+
+    u = TimeFunction(name='u', grid=grid, space_order=2)
+
+    op = Operator(Eq(u.forward, u + 1), dle=('blocking', {'openmp': False,
```
Comment (mloubout) [inline tests/test_autotuner.py]: Why no openmp?

File: tests/test_autotuner.py [hunk 10]
```diff
@@ -296,12 +296,33 @@ def test_multiple_blocking():
...
+                                                          'blocklevels': 2}))
+
+    # 'basic' mode
+    op.apply(time_M=0, autotune='basic')
```
Comment (mloubout) [inline tests/test_autotuner.py]: Haven't seen, but the `runtime` mode still there right?

File: tests/test_dle.py [hunk 11]
```diff
@@ -219,7 +225,7 @@ def test_cache_blocking_edge_cases(shape, blockshape):
 
 
 @pytest.mark.parametrize("shape,blockshape", [
-    ((3, 3), (3, 4)),
+    ((3, 3), (3, 3)),
```
Comment (mloubout) [inline tests/test_dle.py]: Maybe an extra "less-simple" case like `((5,3), (7, 9))` or something like that

File: tests/test_operator.py [hunk 12]
```diff
@@ -442,11 +442,11 @@ def test_staggered_time(self, stagg, ndim):
         Test the "deformed" allocation for staggered functions
         """
         grid = Grid(shape=tuple([11]*ndim))
-        f = TimeFunction(name='f', grid=grid, staggered=stagg)
+        f = TimeFunction(name='f', grid=grid, staggered=stagg, padding=0)
         assert f.data.shape[1:] == tuple(11-i for i in f.staggered[1:])
         # Add a non-staggered field to ensure that the auto-derived
         # dimension size arguments are at maximum
-        g = TimeFunction(name='g', grid=grid)
+        g = TimeFunction(name='g', grid=grid, padding=0)
```
Comment (mloubout) [inline tests/test_operator.py]: wait isn't `padding=0` the default?

File: tests/test_operator.py [hunk 13]
```diff
@@ -1154,65 +1155,108 @@ def test_consistency_coupled_w_ofs(self, exprs, ti0, ti1, ti3):
...
+        eqns = [Eq(u, 1), Eq(v, u.dxr)]
+        op = Operator(eqns)
+        trees = retrieve_iteration_tree(op)
+        assert len(trees) == 2
+        assert trees[0][1].dim is x
+        assert trees[1][1].dim is x
+
+        # Again pretty similar, but with a WAR on `v` -- fission is still
+        # the result of optimization
+        eqns = [Eq(u, v), Eq(v, u.dxl)]
```
Comment (mloubout) [inline tests/test_operator.py]: That's a weird one, not sure from the asserts that it computes second eq first

**Comment by FabioLuporini** [comment]
Merged.

File: devito/dle/blocking_utils.py [hunk 14]
```diff
@@ -58,41 +59,60 @@ def make_blocking(self, iet):
...
                 properties = (PARALLEL,) + ((AFFINE,) if i.is_Affine else ())
-                interb.append(Iteration([], d, d.symbolic_max, properties=properties))
-                # Build Iteration within a block
-                intrab.append(i._rebuild([], limits=(d, d+d.step-1, 1), offsets=(0, 0)))
+
+                # Build Iteration across `level_0` blocks
+                d = BlockDimension(i.dim, name=template % 0)
+                level_0.append(Iteration([], d, d.symbolic_max, properties=properties))
+
+                # Build Iteration across all `level_i` blocks, `i` in (1, self.nlevels]
```
Comment (btrb) [inline devito/dle/blocking_utils.py]: typo at the end? `]`

File: devito/dle/blocking_utils.py [hunk 15]
```diff
@@ -383,30 +405,44 @@ def symbolic_min(self):
...
-            else:
-                # Avoid OOB
-                warning("The specified block size `%s=%d` is bigger than the "
-                        "iteration range; shrinking it to `%s=1`."
-                        % (self.step.name, value, self.step.name))
-                return {self.step.name: 1}
+            return {self.step.name: kwargs.pop(self.step.name)}
+        elif isinstance(self.parent, BlockDimension):
+            # `self` is a BlockDimension within an outer BlockDimension, but
+            # no value supplied -> the sub-block will span the entire block
```
Comment (btrb) [inline devito/dle/blocking_utils.py]: Block?

File: devito/dle/blocking_utils.py [hunk 16]
```diff
@@ -383,30 +405,44 @@ def symbolic_min(self):
...
+            return {self.step.name: args[self.parent.step.name]}
         else:
             value = self._arg_defaults()[self.step.name]
             if value <= args[self.root.max_name] - args[self.root.min_name] + 1:
                 return {self.step.name: value}
             else:
-                # Avoid OOB
+                # Avoid OOB (will end up here only in case of tiny iteration spaces)
```
Comment (btrb) [inline devito/dle/blocking_utils.py]: OOB = out of bounds?

File: devito/dle/rewriters.py [hunk 17]
```diff
@@ -164,6 +154,28 @@ class PlatformRewriter(AbstractRewriter):
     Collection of backend-compiler-specific pragmas.
     """
 
+    _node_parallelizer_type = None
+    """The local-node IET parallelizer. To be specified by subclasses."""
+
+    _default_blocking_levels = 1
+    """
+    Depth of the loop blocking hierarchy. 1 => "blocks", 2 => "blocks" and "sub-blocks",
```
Comment (btrb) [inline devito/dle/rewriters.py]: maybe break line?
```
Depth of the loop blocking hierarchy?
1 => "blocks";
2 => "blocks" and "sub-blocks";
3 => "blocks", "sub-blocks", and "sub-sub-blocks".
```

File: devito/ir/iet/properties.py [hunk 18]
```diff
@@ -51,8 +62,10 @@ class HaloSpotProperty(Tag):
 
 
 USELESS = HaloSpotProperty('useless')
-"""The HaloSpot can be ignored as an halo update at this point would be completely
-useless."""
+"""
+The HaloSpot can be ignored as an halo update at this point would be completely
```
Comment (btrb) [inline devito/ir/iet/properties.py]: an halo?

File: devito/ir/support/basic.py [hunk 19]
```diff
@@ -545,6 +549,24 @@ def is_inplace(self, dim=None):
...
         return self.source.lex_eq(self.sink) and self.is_indep(dim)
 
+    @memoized_meth
+    def is_storage_volatile(self, dims=None):
+        """
+        True if a storage-volatile dependence, False otherwise.
+
+        Examples
+        --------
+        * ``self.function`` is a scalar;
```
Comment (btrb) [inline devito/ir/support/basic.py]: Scalar?

File: devito/types/dense.py [hunk 20]
```diff
@@ -50,17 +51,19 @@ class DiscreteFunction(AbstractCachedFunction, ArgProvider):
...
             self._distributor = self.__distributor_setup__(**kwargs)
 
             # Staggering metadata
             self._staggered = self.__staggered_setup__(**kwargs)
 
+            # Now that *all* __X_setup__ hooks have been called, we can let the
+            # superclass constructor do its job
+            super(DiscreteFunction, self).__init__(*args, **kwargs)
+
+            # There may or may not be a `Grid` attached to the DiscreteFunction
```
Comment (btrb) [inline devito/types/dense.py]: Grid without ``?

File: tests/test_operator.py [hunk 21]
```diff
@@ -1154,65 +1155,108 @@ def test_consistency_coupled_w_ofs(self, exprs, ti0, ti1, ti3):
...
+
+        # Fission as an optimization to increase parallelism
+        eqns = [Eq(u, 1), Eq(v, u.dxl)]
+        op = Operator(eqns)
+        trees = retrieve_iteration_tree(op)
+        assert len(trees) == 2
+        assert trees[0][1].dim is x
+        assert trees[1][1].dim is x
+
+        # Same story as above, but now with a WAR
```
Comment (btrb) [inline tests/test_operator.py]: what's a WAR?

**Pull request merged**